File: whats_new/java/src/main/java/ru/iv/support/dll/Library.java
package ru.iv.support.dll;

import ru.iv.support.Device;
import ru.iv.support.Firmware;
import ru.iv.support.Packet;

import java.io.File;

final class Library {
    private static final boolean LOADED;
    private static final String KEY_LIBRARY_PATH = "java.library.path";
    private static final String USER_DIRECTORY = System.getProperty("user.dir");

    static {
        final StringBuilder libraryPath = new StringBuilder(System.getProperty(KEY_LIBRARY_PATH));
        final File up = new File(USER_DIRECTORY).getParentFile();
        final File build = new File(up, "native" + File.separator + ".build" + File.separator + "bin");
        registerPath(libraryPath, build.getAbsolutePath());
        registerPath(libraryPath, USER_DIRECTORY, "lib");
        registerPath(libraryPath, USER_DIRECTORY, "3rdParty");
        System.setProperty(KEY_LIBRARY_PATH, libraryPath.toString());
        LOADED = loadLibrary();
        Runtime.getRuntime().addShutdownHook(new Thread(Library::destroy));
    }

    private static boolean loadLibrary() {
        try {
            System.loadLibrary("support");
            return init();
        } catch (Exception e) {
            return false;
        }
    }

    static native void register(Callback callback);

    static native void send(Device device, String command, boolean calcHash);

    static native boolean hasDevice(Device device);

    private static native boolean init();

    private static native boolean destroy();

    private static void registerPath(final StringBuilder libraryPath, String... paths) {
        int i = 0;

        libraryPath.append(File.pathSeparator);
        for (String path : paths) {
            if (i > 0) {
                libraryPath.append(File.separator);
            }
            ++i;
            libraryPath.append(path);
        }
    }

    public static void main(String[] args) throws InterruptedException {
        if (LOADED) {
            register(new Callback() {
                @Override
                public void connected(Device device, Firmware firmware) {
                    System.out.println("Connected: " + device + " -> " + firmware);
                }

                @Override
                public void disconnected(Device device, Firmware firmware) {
                    System.out.println("Disconnected: " + device + " -> " + firmware);
                }

                @Override
                public void handle(Device device, Firmware firmware, Packet[] packets, int count) {
                    for (int i = 0; i < count; i++) {
                        System.out.println(packets[i]);
                    }
                }
            });
            System.out.println("Loaded");
            Thread.currentThread().join(5000);
            send(Device.HID, "Q2D\r", true);
        } else {
            System.out.println("Not loaded");
        }
    }
}

File: whats_new/java/src/main/kotlin/Library.kt
import java.util.concurrent.BlockingQueue
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit

data class Packet(val group: Int, val index: Int, val battery: Int, val timeout: Int, val enter: String) {
}

enum class EventType {
    CONNECTED, DISCONNECTED, RECEIVED
}

data class Event(
        val device: Device,
        val firmware: Firmware,
        val type: EventType,
        val packets: Array<Packet> = emptyArray()) {
}

enum class Device {
    HID, FTDI
}

enum class Firmware {
    FW42, FW60
}

interface Callback {
    fun connected(device: Device, firmware: Firmware)
    fun disconnected(device: Device, firmware: Firmware)
    fun handle(device: Device, firmware: Firmware, packets: Array<Packet>, count: Int)
}

open class SystemLibrary private constructor() {
    private val kLibraryPath = "java.library.path";

    init {
        if (loadLibrary() && init()) {
            register(object : Callback {
                override fun connected(device: Device, firmware: Firmware) {
                    events.add(Event(device, firmware, EventType.CONNECTED))
                }

                override fun disconnected(device: Device, firmware: Firmware) {
                    events.add(Event(device, firmware, EventType.DISCONNECTED))
                }

                override fun handle(device: Device, firmware: Firmware, packets: Array<Packet>, count: Int) {
                    events.add(Event(device, firmware, EventType.RECEIVED, packets.sliceArray(IntRange(0, count))));
                }
            })
        }
    }

    open val events: BlockingQueue<Event> = LinkedBlockingQueue(1000)

    open fun device(device: Device): Boolean {
        return hasDevice(device);
    }

    open fun sendCommand(device: Device, command: String): Unit {
        return send(device, command, true);
    }

    private external fun register(cb: Callback);

    private external fun send(device: Device, command: String, calcHash: Boolean)

    private external fun hasDevice(device: Device): Boolean

    private external fun init(): Boolean

    private external fun destroy(): Boolean

    private fun loadLibrary(): Boolean {
        System.loadLibrary("support")
        return true
    }

    private object Holder {
        val INSTANCE = SystemLibrary()
    }

    companion object {
        val default: SystemLibrary by lazy { Holder.INSTANCE }
    }
}

fun main(args: Array<String>) {
    val events: BlockingQueue<Event> = SystemLibrary.default.events;
    val runnable = Runnable {
        while (true) {
            val event = try {
                events.poll(100, TimeUnit.MILLISECONDS)
            } catch(e: InterruptedException) {
                null
            }
            if (event != null) {
                System.out.println(event)
            }
        }
    };
    val thread = Thread(runnable)
    thread.start()
    thread.join()
}
